README.md
# react-native-sqlite-storage

SQLite3 Native Plugin for React Native for both Android (Classic and Native) and iOS

Foundation of this library is based on Chris Brody's Cordova SQLite plugin.

Features:

1. iOS and Android supported via identical JavaScript API.
2. Android in pure Java and Native modes
3. SQL transactions
4. JavaScript interface via plain callbacks or Promises.
5. Pre-populated SQLite database import from application bundle and sandbox

There are sample apps provided in test directory that can be used in with the AwesomeProject generated by React Native. All you have to do is to copy one of those files into your AwesomeProject replacing index.ios.js.

Please let me know your projects that use these SQLite React Native modules. I will list them in the reference section. If there are any features that you think would benefit this library please post them.

The library has been tested with React 16.2 (and earlier) and XCode 7,8,9 - it works fine out of the box without any need for tweaks or code changes. For XCode 7,8 vs. XCode 6 the only difference is that sqlite ios library name suffix is tbd instead of dylib.

Version 3.2 is the first version compatible with RN 0.40.

# How to use (iOS):

#### Step 1. Install Dependencies

##### With CocoaPods:

This command should be executed in the root directory of your RN project

```shell
npm install --save react-native-sqlite-storage
```

Then add this to your Podfile which should be located inside the ios project subdirectory

```ruby
pod 'React', :path => '../node_modules/react-native'
pod 'react-native-sqlite-storage', :path => '../node_modules/react-native-sqlite-storage'
```

Or use the sample Podfile included in the package by copying it over to ios subdirectory and replacing AwesomeProject inside of it with the name of your RN project.

Refresh the Pods installation

```ruby
pod install
```

OR

```ruby
pod update
```

Done, skip to Step 2.

##### Without CocoaPods:

This command should be executed in the root directory of your RN project

```shell
npm install --save react-native-sqlite-storage
rnpm link
```

rnpm and xcode are dependencies of this project and should get installed with the module but in case there are issue running rnpm link and rnpm/xcode are not already installed you can try to install it globally as follows:

```shell
npm -g install rnpm xcode
```

After linking project should like this:

![alt tag](instructions/after-rnpm.png)

#### Step 1a. If rnpm link does not work for you you can try manually linking according to the instructions below:

##### Drag the SQLite Xcode project as a dependency project into your React Native XCode project

![alt tag](https://raw.github.com/andpor/react-native-sqlite-storage/master/instructions/libs.png)

##### XCode SQLite libraries dependency set up

Add libSQLite.a (from Workspace location) to the required Libraries and Frameworks. Also add sqlite3.0.tbd (XCode 7) or libsqlite3.0.dylib (XCode 6 and earlier) in the same fashion using Required Libraries view (Do not just add them manually as the build paths will not be properly set)

![alt tag](https://raw.github.com/andpor/react-native-sqlite-storage/master/instructions/addlibs.png)

#### Step 2. Application JavaScript require

Add var SQLite = require('react-native-sqlite-storage') to your index.ios.js

![alt tag](instructions/require.png)

#### Step 3. Write application JavaScript code using the SQLite plugin

Add JS application code to use SQLite API in your index.ios.js etc. Here is some sample code. For full working example see test/index.ios.callback.js. Please note that Promise based API is now supported as well with full examples in the working React Native app under test/index.ios.promise.js

```javascript
errorCB(err) {
  console.log("SQL Error: " + err);
},

successCB() {
  console.log("SQL executed fine");
},

openCB() {
  console.log("Database OPENED");
},

var db = SQLite.openDatabase("test.db", "1.0", "Test Database", 200000, openCB, errorCB);
db.transaction((tx) => {
  tx.executeSql('SELECT * FROM Employees a, Departments b WHERE a.department = b.department_id', [], (tx, results) => {
      console.log("Query completed");

      // Get rows with Web SQL Database spec compliance.

      var len = results.rows.length;
      for (let i = 0; i < len; i++) {
        let row = results.rows.item(i);
        console.log(`Employee name: ${row.name}, Dept Name: ${row.deptName}`);
      }

      // Alternatively, you can use the non-standard raw method.

      /*
        let rows = results.rows.raw(); // shallow copy of rows Array

        rows.map(row => console.log(`Employee name: ${row.name}, Dept Name: ${row.deptName}`));
      */
    });
});
```

# How to use (Android):

#### Step 1 - NPM Install

```shell
npm install --save react-native-sqlite-storage
```

#### Step 2 - Update Gradle Settings (located under Gradle Settings in Project Panel)

```gradle
// file: android/settings.gradle
...

include ':react-native-sqlite-storage'
project(':react-native-sqlite-storage').projectDir = new File(rootProject.projectDir, '../node_modules/react-native-sqlite-storage/src/android')
```

#### Step 3 - Update app module Gradle Build script (located under Gradle Settings in Project Panel)

```gradle
// file: android/app/build.gradle
...

dependencies {
    ...
    compile project(':react-native-sqlite-storage')
}
```

#### Step 4 - Register React Package (this should work on React version but if it does not , try the ReactActivity based approach. Note: for version 3.0.0 and below you would have to pass in the instance of your Activity to the SQLitePluginPackage constructor

```java
...
import org.pgsqlite.SQLitePluginPackage;

public class MainActivity extends Activity implements DefaultHardwareBackBtnHandler {

    private ReactInstanceManager mReactInstanceManager;
    private ReactRootView mReactRootView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mReactRootView = new ReactRootView(this);
        mReactInstanceManager = ReactInstanceManager.builder()
                .setApplication(getApplication())
                .setBundleAssetName("index.android.bundle")  // this is dependant on how you name you JS files, example assumes index.android.js
                .setJSMainModuleName("index.android")        // this is dependant on how you name you JS files, example assumes index.android.js
                .addPackage(new MainReactPackage())
                .addPackage(new SQLitePluginPackage())       // register SQLite Plugin here
                .setUseDeveloperSupport(BuildConfig.DEBUG)
                .setInitialLifecycleState(LifecycleState.RESUMED)
                .build();
        mReactRootView.startReactApplication(mReactInstanceManager, "AwesomeProject", null); //change "AwesomeProject" to name of your app
        setContentView(mReactRootView);
    }
...
```

Alternative approach on newer versions of React Native (0.18+). Note: for version 3.0.0 and below you would have to pass in the instance of your Activity to the SQLitePluginPackage constructor

```java
import org.pgsqlite.SQLitePluginPackage;

public class MainApplication extends Application implements ReactApplication {
  ......

  /**
   * A list of packages used by the app. If the app uses additional views
   * or modules besides the default ones, add more packages here.
   */
    @Override
    protected List<ReactPackage> getPackages() {
      return Arrays.<ReactPackage>asList(
        new SQLitePluginPackage(),   // register SQLite Plugin here
        new MainReactPackage())
    }
}
```

#### Step 5 - Require and use in Javascript - see full examples (callbacks and Promise) in test directory.

```js
// file: index.android.js

var React = require('react-native');
var SQLite = require('react-native-sqlite-storage')
...
```

## Setting up your project to import a pre-populated SQLite database from application for iOS

#### Step 1 - Create 'www' folder.

Create a folder called 'www' (yes must be called precisely that else things won't work) in the project folder via Finder

#### Step 2 - Create the database file

Copy/paste your pre-populated database file into the 'www' folder. Give it the same name you are going to use in openDatabase call in your application

#### Step 3 - Add file to project

in XCode, right click on the main folder and select Add Files to 'your project name'

![alt tag](https://raw.github.com/andpor/react-native-sqlite-storage/master/instructions/addFilesToProject.png)

#### Step 4 - Choose files to add

In the Add Files dialog, navigate to the 'www' directory you created in Step 1, select it, make sure you check the option to Create Folder Reference

![alt tag](https://raw.github.com/andpor/react-native-sqlite-storage/master/instructions/addFilesToProjectSelect.png)

#### Step 5 - Verify project structure

Ensure your project structure after previous steps are executed looks like this

![alt tag](https://raw.github.com/andpor/react-native-sqlite-storage/master/instructions/projectStructureAfter.png)

### Step 6 - Adjust openDatabase call

Modify you openDatabase call in your application adding createFromLocation param. If you named your database file in step 2 'testDB' the openDatabase call should look like something like this:

```js
  ...
  1.SQLite.openDatabase({name : "testDB", createFromLocation : 1}, okCallback,errorCallback);
  // default - if your folder is called www and data file is named the same as the dbName - testDB in this example
  2.SQLite.openDatabase({name : "testDB", createFromLocation : "~data/mydbfile.sqlite"}, okCallback,errorCallback);
  // if your folder is called data rather than www or your filename does not match the name of the db
  3.SQLite.openDatabase({name : "testDB", createFromLocation : "/data/mydbfile.sqlite"}, okCallback,errorCallback);
  // if your folder is not in app bundle but in app sandbox i.e. downloaded from some remote location.
  ...
```

For Android, the www directory is always relative to the assets directory for the app: src/main/assets

Enjoy!

## Opening a database

Opening a database is slightly different between iOS and Android. Where as on Android the location of the database file is fixed, there are three choices of where the database file can be located on iOS. The 'location' parameter you provide to openDatabase call indicated where you would like the file to be created. This parameter is neglected on Android.

WARNING: the default location on iOS has changed in version 3.0.0 - it is now a no-sync location as mandated by Apple so the release is backward incompatible.

To open a database in default no-sync location (affects iOS _only_)::

```js
SQLite.openDatabase({ name: "my.db", location: "default" }, successcb, errorcb);
```

To specify a different location (affects iOS _only_):

```js
SQLite.openDatabase({ name: "my.db", location: "Library" }, successcb, errorcb);
```

where the `location` option may be set to one of the following choices:

- `default`: `Library/LocalDatabase` subdirectory - _NOT_ visible to iTunes and _NOT_ backed up by iCloud
- `Library`: `Library` subdirectory - backed up by iCloud, _NOT_ visible to iTunes
- `Documents`: `Documents` subdirectory - visible to iTunes and backed up by iCloud

The original webSql style openDatabase still works and the location will implicitly default to 'default' option:

```js
SQLite.openDatabase("myDatabase.db", "1.0", "Demo", -1);
```

## Importing a pre-populated database.

You can import an existing - prepopulated database file into your application. Depending on your instructions in openDatabase call, the sqlite-storage will look at different places to locate you pre-populated database file.

Use this flavor of openDatabase call, if your folder is called www and data file is named the same as the dbName - testDB in this example

```js
SQLite.openDatabase(
  { name: "testDB", createFromLocation: 1 },
  okCallback,
  errorCallback
);
```

Use this flavor of openDatabase call if your folder is called data rather than www or your filename does not match the name of the db. In this case db is named testDB but the file is mydbfile.sqlite which is located in a data subdirectory of www

```js
SQLite.openDatabase(
  { name: "testDB", createFromLocation: "~data/mydbfile.sqlite" },
  okCallback,
  errorCallback
);
```

Use this flavor of openDatabase call if your folder is not in application bundle but in app sandbox i.e. downloaded from some remote location. In this case the source file is located in data subdirectory of Documents location (iOS) or FilesDir (Android).

```js
SQLite.openDatabase(
  { name: "testDB", createFromLocation: "/data/mydbfile.sqlite" },
  okCallback,
  errorCallback
);
```

## Additional options for pre-populated database file

You can provide additional instructions to sqlite-storage to tell it how to handle your pre-populated database file. By default, the source file is copied over to the internal location which works in most cases but sometimes this is not really an option particularly when the source db file is large. In such situations you can tell sqlite-storage you do not want to copy the file but rather use it in read-only fashion via direct access. You accomplish this by providing an additional optional readOnly parameter to openDatabase call

```js
SQLite.openDatabase(
  {
    name: "testDB",
    readOnly: true,
    createFromLocation: "/data/mydbfile.sqlite"
  },
  okCallback,
  errorCallback
);
```

Note that in this case, the source db file will be open in read-only mode and no updates will be allowed. You cannot delete a database that was open with readOnly option. For Android, the read only option works with pre-populated db files located in FilesDir directory because all other assets are never physically located on the file system but rather read directly from the app bundle.

## Attaching another database

Sqlite3 offers the capability to attach another database to an existing database-instance, i.e. for making cross database JOINs available.
This feature allows to SELECT and JOIN tables over multiple databases with only one statement and only one database connection.
To archieve this, you need to open both databases and to call the attach()-method of the destination (or master) -database to the other ones.

```js
let dbMaster, dbSecond;

dbSecond = SQLite.openDatabase(
  { name: "second" },
  db => {
    dbMaster = SQLite.openDatabase(
      { name: "master" },
      db => {
        dbMaster.attach(
          "second",
          "second",
          () => console.log("Database attached successfully"),
          () => console.log("ERROR")
        );
      },
      err => console.log("Error on opening database 'master'", err)
    );
  },
  err => console.log("Error on opening database 'second'", err)
);
```

The first argument of attach() is the name of the database, which is used in SQLite.openDatabase(). The second argument is the alias, that is used to query on tables of the attached database.

The following statement would select data from the master database and include the "second"-database within a simple SELECT/JOIN-statement:

```sql
SELECT * FROM user INNER JOIN second.subscriptions s ON s.user_id = user.id
```

To detach a database, just use the detach()-method:

```js
dbMaster.detach("second", successCallback, errorCallback);
```

For sure, their is also Promise-support available for attach() and detach(), as shown in the example-application under the
directory "examples".

# Original Cordova SQLite Bindings from Chris Brody

https://github.com/litehelpers/Cordova-sqlite-storage

The issues and limitations for the actual SQLite can be found on this site.

## Issues

1. Android binds all numeric SQL input values to double. This is due to the underlying React Native limitation where only a Numeric type is available on the interface point making it ambiguous to distinguish integers from doubles. Once I figure out the proper way to do this I will update the codebase [(Issue #4141)](https://github.com/facebook/react-native/issues/4141).


CHANGELOG.md
# Version History

v3.3.6

1.  Fix INTEGER column value overflow.

v3.3.5

1.  All JSON conversions in Android modules have been eliminated in favor of direct React Native interfaces. This should provide siginificant performance improvement.
2.  Main Queue Warning issue fixed in iOS.
3.  Exception handling in logic processing asset imports have been fixed and error codes are properly propagated to JS callabcks.
4.  Examples have been revamped and library was tested with XCode 9 and Android Studio 3.1.1 (Gradle 2.2)

v3.3.4

1.  New version of native binaries compile with latest sqlite version 3.20.1 supporting pcre extension to enable REGEXP function for Android. (https://github.com/andpor/react-native-sqlite-storage/pull/205)
2.  Fixes Xcode warning for potentially insecure string. (https://github.com/andpor/react-native-sqlite-storage/pull/199)
3.  Remove createJSModules @ovveride marker - RN 0.47 compatibility. (https://github.com/andpor/react-native-sqlite-storage/pull/188)
4.  Podfile add macOS support (https://github.com/andpor/react-native-sqlite-storage/pull/179)
5.  instructions directory added on npmignore. (https://github.com/andpor/react-native-sqlite-storage/pull/174)

v3.3.3

1.  Fix for the db location string [Issue #172](https://github.com/andpor/react-native-sqlite-storage/issues/172)
2.  #define in iOS casebase for SQLCIPHER. If you include this #define in your main project settings, this library will pick up appropriate key handling code automatically.

v3.3.2

1.  Yoga import fix

v3.3.1

1.  Comment of SQLCipher code in iOS implementation as a quick fix. [Issue #155](https://github.com/andpor/react-native-sqlite-storage/issues/155)

v3.3.0 (Extended thanks to dryganets for his significant contributions to this release)

1.  Access to MutableDictonary openDBs properly synchronized [PR #130](https://github.com/andpor/react-native-sqlite-storage/pull/130)
2.  Database attach flow fixed. Threading model fix in order to have all queries executed in the same order [PR #131](https://github.com/andpor/react-native-sqlite-storage/pull/131)
3.  All statements and queries are closed in finally statements in order to fix SQLiteCipher corner case crashes [PR #132](https://github.com/andpor/react-native-sqlite-storage/pull/132)
4.  Minor style fix in index.ios.callback.js and index.ios.promise.js [PR #136](https://github.com/andpor/react-native-sqlite-storage/pull/136)
5.  Fix determination logic for opened db [PR #139](https://github.com/andpor/react-native-sqlite-storage/pull/139)
6.  Clean up in lib/sqlite.core.js [PR #138](https://github.com/andpor/react-native-sqlite-storage/pull/138)
7.  Production grade logging for the Android plugin [PR #137](https://github.com/andpor/react-native-sqlite-storage/pull/137)
8.  Remove pre-honeycomb workaround code in Android that was causing issues in SQL Cipher [PR #147](https://github.com/andpor/react-native-sqlite-storage/pull/147)
9.  Fix broken Markdown headings [PR #153](https://github.com/andpor/react-native-sqlite-storage/pull/153)
10. Drop usage of the dead rnpm repository [PR #148](https://github.com/andpor/react-native-sqlite-storage/pull/148)

v3.2.2

1.  Corrects the CocoaPods based development set-up instructions and includes sample Podfile. [Issue #125](https://github.com/andpor/react-native-sqlite-storage/issues/125)

v3.2.1

1.  Sample apps in test directory adjusted for React Native 0.41 and plugability in AwesomeProject. [Issue #120](https://github.com/andpor/react-native-sqlite-storage/issues/120)

v3.2.0

1.  This is a backward incompatible release with baseline React Native 0.40 support.
2.  [React Native 0.40 compatibility fixes](https://github.com/andpor/react-native-sqlite-storage/pull/110) - thanks K-Leon for this contribution

v3.1.3

1.  Add support for ATTACH (thanks to itinance for this contribution)
2.  Example applications are now hosted in separate repo [react-native-sqlite-examples](https://github.com/andpor/react-native-sqlite-storage-examples)

v3.1.2

1.  Add support for CocoaPods (thanks to JAStanton for this contribution)
2.  Set base iOS build to 8.0

v3.1.1

1.  Fix for Cordova issue #517: reject ALTER, REINDEX and REPLACE operations in readTransactions
2.  Stop remaining transaction callback in case of an error with no error handler returning false

v3.1.0

1.  Backward incompatible change. Boolean params will now be converted and stored as int type, 0 and 1, in compliance with SQLite specifications. Issue [#63](https://github.com/andpor/react-native-sqlite-storage/issues/63)
2.  Database decoupled from the Activity lifecycle on Android. With this change, the database will not be closed without explicit instructions to close it. SQLitePluginPackage constructor change. Pull Request [#62](https://github.com/andpor/react-native-sqlite-storage/pull/62)
3.  Correct handling for executeSql with object as sql value (solves a possible crash on iOS)
4.  Backfill cordova-sqlite-storage fix - readTransaction allows modification in case of extra semicolon before SQL. Issue [#460](https://github.com/litehelpers/Cordova-sqlite-storage/issues/460)

v3.0.0

1.  Default location changes for iOS for App Store iCloud compliance - backward incompatible release. Default now is no-sync location instead of docs.
2.  Ability to point to read-only db file in app bundle directly without requiring it to be copied elsewhere.
3.  Check if db is open before throwing an exception (triggered in android lock workaround)
4.  Fix for issue #57. Can't find variable: Blob

v2.1.6

1.  rnpm linking for iOS - contributed by @clozr
2.  Backfill Cordova read transaction bug fix.

v2.1.5

1.  Allow retrieval of pre-populated db files from user defined locations in application bundle as well as the sandbox.
2.  Implement Activity lifecycle mgmt in Android native
3.  Fix issue [#37](https://github.com/andpor/react-native-sqlite-storage/issues/37) - Int Column type value overflow

test/index.ios.promise.js
/*
 * sqlite.ios.promise.js
 *
 * Created by Andrzej Porebski on 10/29/15.
 * Copyright (c) 2015 Andrzej Porebski.
 *
 * Test App using Promise for react-naive-sqlite-storage
 *
 * This library is available under the terms of the MIT License (2008).
 * See http://opensource.org/licenses/alphabetical for full text.
 */
"use strict";

import React, { Component } from "react";
import { AppRegistry, StyleSheet, Text, View, ListView } from "react-native";

import SQLite from "react-native-sqlite-storage";
SQLite.DEBUG(true);
SQLite.enablePromise(true);

const database_name = "Test.db";
const database_version = "1.0";
const database_displayname = "SQLite Test Database";
const database_size = 200000;
let db;

class SQLiteDemo extends Component {
  constructor() {
    super();
    this.progress = [];
    this.state = {
      progress: [],
      ds: new ListView.DataSource({
        rowHasChanged: (r1, r2) => r1 !== r2
      })
    };
  }

  updateProgress = (text, resetState) => {
    let progress = [];
    if (!resetState) {
      progress = [...this.progress];
    }
    progress.push(text);
    this.progress = progress;
    this.setState({
      progress
    });
  };

  componentWillUnmount() {
    this.closeDatabase();
  }

  errorCB = err => {
    console.log("error: ", err);
    this.updateProgress("Error " + (err.message || err));
  };

  populateDatabase = db => {
    this.updateProgress("Database integrity check");
    db.executeSql("SELECT 1 FROM Version LIMIT 1")
      .then(() => {
        this.updateProgress("Database is ready ... executing query ...");
        db.transaction(this.queryEmployees).then(() => {
          this.updateProgress("Processing completed");
        });
      })
      .catch(error => {
        console.log("Received error: ", error);
        this.updateProgress("Database not yet ready ... populating data");
        db.transaction(this.populateDB).then(() => {
          this.updateProgress("Database populated ... executing query ...");
          db.transaction(this.queryEmployees).then(result => {
            console.log("Transaction is now finished");
            this.updateProgress("Processing completed");
            this.closeDatabase();
          });
        });
      });
  };

  populateDB = tx => {
    this.updateProgress("Executing DROP stmts");

    tx.executeSql("DROP TABLE IF EXISTS Employees;");
    tx.executeSql("DROP TABLE IF EXISTS Offices;");
    tx.executeSql("DROP TABLE IF EXISTS Departments;");

    this.updateProgress("Executing CREATE stmts");

    tx.executeSql(
      "CREATE TABLE IF NOT EXISTS Version( " +
        "version_id INTEGER PRIMARY KEY NOT NULL); "
    ).catch(error => {
      this.errorCB(error);
    });

    tx.executeSql(
      "CREATE TABLE IF NOT EXISTS Departments( " +
        "department_id INTEGER PRIMARY KEY NOT NULL, " +
        "name VARCHAR(30) ); "
    ).catch(error => {
      this.errorCB(error);
    });

    tx.executeSql(
      "CREATE TABLE IF NOT EXISTS Offices( " +
        "office_id INTEGER PRIMARY KEY NOT NULL, " +
        "name VARCHAR(20), " +
        "longtitude FLOAT, " +
        "latitude FLOAT ) ; "
    ).catch(error => {
      this.errorCB(error);
    });

    tx.executeSql(
      "CREATE TABLE IF NOT EXISTS Employees( " +
        "employe_id INTEGER PRIMARY KEY NOT NULL, " +
        "name VARCHAR(55), " +
        "office INTEGER, " +
        "department INTEGER, " +
        "FOREIGN KEY ( office ) REFERENCES Offices ( office_id ) " +
        "FOREIGN KEY ( department ) REFERENCES Departments ( department_id ));"
    ).catch(error => {
      this.errorCB(error);
    });

    this.updateProgress("Executing INSERT stmts");

    tx.executeSql('INSERT INTO Departments (name) VALUES ("Client Services");');
    tx.executeSql(
      'INSERT INTO Departments (name) VALUES ("Investor Services");'
    );
    tx.executeSql('INSERT INTO Departments (name) VALUES ("Shipping");');
    tx.executeSql('INSERT INTO Departments (name) VALUES ("Direct Sales");');

    tx.executeSql(
      'INSERT INTO Offices (name, longtitude, latitude) VALUES ("Denver", 59.8,  34.1);'
    );
    tx.executeSql(
      'INSERT INTO Offices (name, longtitude, latitude) VALUES ("Warsaw", 15.7, 54.1);'
    );
    tx.executeSql(
      'INSERT INTO Offices (name, longtitude, latitude) VALUES ("Berlin", 35.3, 12.1);'
    );
    tx.executeSql(
      'INSERT INTO Offices (name, longtitude, latitude) VALUES ("Paris", 10.7, 14.1);'
    );

    tx.executeSql(
      'INSERT INTO Employees (name, office, department) VALUES ("Sylvester Stallone", 2,  4);'
    );
    tx.executeSql(
      'INSERT INTO Employees (name, office, department) VALUES ("Elvis Presley", 2, 4);'
    );
    tx.executeSql(
      'INSERT INTO Employees (name, office, department) VALUES ("Leslie Nelson", 3,  4);'
    );
    tx.executeSql(
      'INSERT INTO Employees (name, office, department) VALUES ("Fidel Castro", 3, 3);'
    );
    tx.executeSql(
      'INSERT INTO Employees (name, office, department) VALUES ("Bill Clinton", 1, 3);'
    );
    tx.executeSql(
      'INSERT INTO Employees (name, office, department) VALUES ("Margaret Thatcher", 1, 3);'
    );
    tx.executeSql(
      'INSERT INTO Employees (name, office, department) VALUES ("Donald Trump", 1, 3);'
    );
    tx.executeSql(
      'INSERT INTO Employees (name, office, department) VALUES ("Dr DRE", 2, 2);'
    );
    tx.executeSql(
      'INSERT INTO Employees (name, office, department) VALUES ("Samantha Fox", 2, 1);'
    );
    console.log("all config SQL done");
  };

  queryEmployees = tx => {
    console.log("Executing employee query");
    tx.executeSql(
      "SELECT a.name, b.name as deptName FROM Employees a, Departments b WHERE a.department = b.department_id and a.department=?",
      [3]
    )
      .then(([tx, results]) => {
        this.updateProgress("Query completed");
        var len = results.rows.length;
        for (let i = 0; i < len; i++) {
          let row = results.rows.item(i);
          this.updateProgress(
            `Empl Name: ${row.name}, Dept Name: ${row.deptName}`
          );
        }
      })
      .catch(error => {
        console.log(error);
      });
  };

  loadAndQueryDB = () => {
    this.updateProgress("Plugin integrity check ...");
    SQLite.echoTest()
      .then(() => {
        this.updateProgress("Integrity check passed ...");
        this.updateProgress("Opening database ...");
        SQLite.openDatabase({
          name: "test5.db",
          createFromLocation: "~/db/andrew.db"
        })
          .then(DB => {
            db = DB;
            this.updateProgress("Database OPEN");
            this.populateDatabase(DB);
          })
          .catch(error => {
            console.log(error);
          });
      })
      .catch(error => {
        this.updateProgress("echoTest failed - plugin not functional");
      });
  };

  closeDatabase = () => {
    if (db) {
      console.log("Closing database ...");
      this.updateProgress("Closing DB");
      db.close()
        .then(status => {
          this.updateProgress("Database CLOSED");
        })
        .catch(error => {
          this.errorCB(error);
        });
    } else {
      this.updateProgress("Database was not OPENED");
    }
  };

  deleteDatabase = () => {
    this.updateProgress("Deleting database");
    SQLite.deleteDatabase(database_name)
      .then(() => {
        console.log("Database DELETED");
        this.updateProgress("Database DELETED");
      })
      .catch(error => {
        this.errorCB(error);
      });
  };

  runDemo = () => {
    console.log("running");
    this.updateProgress("Starting SQLite Promise Demo", true);
    this.loadAndQueryDB();
  };

  renderProgressEntry = entry => {
    return (
      <View style={listStyles.li}>
        <View>
          <Text style={listStyles.liText}>{entry}</Text>
        </View>
      </View>
    );
  };

  render() {
    let ds = new ListView.DataSource({
      rowHasChanged: (row1, row2) => row1 !== row2
    });
    return (
      <View style={styles.mainContainer}>
        <View style={styles.toolbar}>
          <Text style={styles.toolbarButton} onPress={this.runDemo}>
            Run Demo
          </Text>
          <Text style={styles.toolbarButton} onPress={this.closeDatabase}>
            Close DB
          </Text>
          <Text style={styles.toolbarButton} onPress={this.deleteDatabase}>
            Delete DB
          </Text>
        </View>
        <ListView
          enableEmptySections={true}
          dataSource={this.state.ds.cloneWithRows(this.state.progress)}
          renderRow={this.renderProgressEntry}
          style={listStyles.liContainer}
        />
      </View>
    );
  }
}

var listStyles = StyleSheet.create({
  li: {
    borderBottomColor: "#c8c7cc",
    borderBottomWidth: 0.5,
    paddingTop: 15,
    paddingRight: 15,
    paddingBottom: 15
  },
  liContainer: {
    backgroundColor: "#fff",
    flex: 1,
    paddingLeft: 15
  },
  liIndent: {
    flex: 1
  },
  liText: {
    color: "#333",
    fontSize: 17,
    fontWeight: "400",
    marginBottom: -3.5,
    marginTop: -3.5
  }
});

var styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    backgroundColor: "#F5FCFF"
  },
  welcome: {
    fontSize: 20,
    textAlign: "center",
    margin: 10
  },
  instructions: {
    textAlign: "center",
    color: "#333333",
    marginBottom: 5
  },
  toolbar: {
    backgroundColor: "#51c04d",
    paddingTop: 30,
    paddingBottom: 10,
    flexDirection: "row"
  },
  toolbarButton: {
    color: "blue",
    textAlign: "center",
    flex: 1
  },
  mainContainer: {
    flex: 1
  }
});

AppRegistry.registerComponent("AwesomeProject", () => SQLiteDemo);


scripts/rnpm-postlink.js
#!/usr/bin/env node

"use strict";

require("react-native/setupBabel")();
var fs = require("fs");
var cli = require("react-native/local-cli/core");
var xcode = require("xcode");

// The current working directory should be project root of the app that is linking react-native-sqlite.
var config = cli.getProjectConfig();

if (config.ios) {
  var pbxproj = config.ios.pbxprojPath;
  var project = xcode.project(pbxproj).parseSync();
  var target = project.getFirstTarget().uuid;

  // Create a Frameworks group if necessary.
  if (!project.pbxGroupByName("Frameworks")) {
    var group = project.pbxCreateGroup("Frameworks", '""');
    var mainGroup = project.getFirstProject().firstProject.mainGroup;

    project.getPBXGroupByKey(mainGroup).children.push({
      value: group,
      comment: "Frameworks"
    });
  }

  ["libsqlite3.0"].forEach(function(name) {
    project.addFramework("usr/lib/" + name + ".tbd", {
      lastKnownFileType: "sourcecode.text-based-dylib-definition",
      sourceTree: "SDKROOT",
      target: target
    });
  });

  fs.writeFileSync(pbxproj, project.writeSync());
}


sqlite.js
/*
 * sqlite.js
 *
 * Created by Andrzej Porebski on 10/29/15.
 * Copyright (c) 2015-16 Andrzej Porebski.
 *
 * This library is available under the terms of the MIT License (2008).
 * See http://opensource.org/licenses/alphabetical for full text.
 */
var plugin = require("./lib/sqlite.core.js");
var { SQLiteFactory } = plugin;

var config = [
  // meaning: [returnValueExpected,prototype,fn,argsNeedPadding,reverseCallbacks,rejectOnError]

  [false, "SQLitePlugin", "transaction", false, true, true],
  [false, "SQLitePlugin", "readTransaction", false, true, true],
  [false, "SQLitePlugin", "close", false, false, true],
  [false, "SQLitePlugin", "executeSql", true, false, true],
  [false, "SQLitePlugin", "sqlBatch", false, false, true],
  [false, "SQLitePlugin", "attach", true, false, true],
  [false, "SQLitePlugin", "detach", false, false, true],
  [false, "SQLitePluginTransaction", "executeSql", true, false, false],
  [false, "SQLiteFactory", "deleteDatabase", false, false, true],
  [true, "SQLiteFactory", "openDatabase", false, false, true],
  [false, "SQLiteFactory", "echoTest", false, false, true]
];

var originalFns = {};
config.forEach(entry => {
  let [returnValueExpected, prototype, fn] = entry;
  let originalFn = plugin[prototype].prototype[fn];
  originalFns[prototype + "." + fn] = originalFn;
});

function enablePromiseRuntime(enable) {
  if (enable) {
    createPromiseRuntime();
  } else {
    createCallbackRuntime();
  }
}
function createCallbackRuntime() {
  config.forEach(entry => {
    let [
      returnValueExpected,
      prototype,
      fn,
      argsNeedPadding,
      reverseCallbacks,
      rejectOnError
    ] = entry;
    plugin[prototype].prototype[fn] = originalFns[prototype + "." + fn];
  });
  console.log("Callback based runtime ready");
}
function createPromiseRuntime() {
  config.forEach(entry => {
    let [
      returnValueExpected,
      prototype,
      fn,
      argsNeedPadding,
      reverseCallbacks,
      rejectOnError
    ] = entry;
    let originalFn = plugin[prototype].prototype[fn];
    plugin[prototype].prototype[fn] = function(...args) {
      if (argsNeedPadding && args.length == 1) {
        args.push([]);
      }
      var promise = new Promise((resolve, reject) => {
        let success = function(...args) {
          if (!returnValueExpected) {
            return resolve(args);
          }
        };
        let error = function(err) {
          console.log("error: ", fn, ...args, arguments);
          if (rejectOnError) {
            reject(err);
          }
          return false;
        };
        var retValue = originalFn.call(
          this,
          ...args,
          reverseCallbacks ? error : success,
          reverseCallbacks ? success : error
        );
        if (returnValueExpected) {
          return resolve(retValue);
        }
      });

      return promise;
    };
  });
  console.log("Promise based runtime ready");
}
SQLiteFactory.prototype.enablePromise = enablePromiseRuntime;

module.exports = new SQLiteFactory();
